README.md
# Manchu-Chinese-Dictionary_Query-Tool

基本使用方法：
運行dict.exe，輸入要查找的單詞。拉丁轉寫遵循《滿漢大辭典》的規定。

功能：
1、顯示單詞所在頁數（pdf文件頁數，非詞典內頁數）。由於下列原因，該頁數不一定完全準確：單詞可能不存在，在上一頁中也有該單詞的詞條內容，索引錯誤，或詞典內容錯誤。
2、使用第三方的pdf瀏覽軟件打開該單詞所在頁。（請使用經去除重複頁碼處理的《滿漢大辭典》pdf文件
3、保存查詢記錄（單詞，頁數，查詢時間）

請使用settings.ini開啓或關閉功能2與功能3。若無法正確設置使用第三方pdf瀏覽軟件，則關閉功能2。設置方法詳見settings.ini文件內註釋。

By Yize Hu, yizehu@gmail.com, 31/5/2014. Written with python 2.7.


history.txt
dergi,308,27/05/2014 17:50:06
abka,125,27/05/2014 17:50:15
zhakuun,972,27/05/2014 21:22:19
abka,125,28/05/2014 21:28:58
zhulergi,1006,28/05/2014 21:30:35
dergi,689,30/05/2014 20:49:23
dergi,705,30/05/2014 20:51:13
julergi,894,30/05/2014 20:51:23
fejergi,1053,30/05/2014 20:51:35
dergi,705,30/05/2014 20:53:54
dergi,705,30/05/2014 20:56:48
dergi,705,30/05/2014 20:58:36
abka,44,31/05/2014 09:37:50
sengguwembi,502,31/05/2014 09:38:39
alambi,53,31/05/2014 09:39:51
arambi,85,31/05/2014 09:40:04
daicing,670,31/05/2014 23:34:36
fukjin,1075,31/05/2014 23:36:01
doro,717,31/05/2014 23:36:37
neihe,254,31/05/2014 23:37:43
bodogon,453,31/05/2014 23:38:47
bithe,437,31/05/2014 23:39:13
daicing,670,01/06/2014 13:18:45
daicing,670,01/06/2014 15:25:25
fukjin,1075,01/06/2014 15:26:25
doro,717,01/06/2014 15:27:04
neihe,254,01/06/2014 15:28:00
neimbi,254,01/06/2014 15:29:12
bodogon,453,01/06/2014 15:30:04
hada,351,01/06/2014 15:35:06
dade,683,01/06/2014 15:35:36
hvlun,387,01/06/2014 15:36:26
bimbi,440,01/06/2014 15:39:01
hala,352,01/06/2014 15:39:35
sekiyen,509,01/06/2014 16:39:13
gebu,960,01/06/2014 16:39:38
nacibulu,250,01/06/2014 16:40:22
ahvngga,39,01/06/2014 16:54:31
jacin,862,01/06/2014 16:55:36
tukiyembi,655,01/06/2014 17:01:46
junda,889,01/06/2014 17:05:18
fonda,1066,01/06/2014 17:06:11
gunehe,994,01/06/2014 17:07:32
dahanha,673,01/06/2014 17:10:25
dahambi,674,01/06/2014 17:10:55
fusihvn,1077,01/06/2014 17:11:41
gisun,978,01/06/2014 17:12:13
gaibumbi,309,01/06/2014 17:12:42
urse,240,01/06/2014 17:15:24
jubobumbi,892,01/06/2014 17:19:31
dube,726,01/06/2014 17:22:42
efujembi,147,01/06/2014 17:24:14
isibumbi,157,01/06/2014 17:24:46
kai,286,01/06/2014 17:25:42
sahaliyan,489,01/06/2014 17:44:16
dalin,684,01/06/2014 17:45:49
tembi,620,01/06/2014 17:46:18
gurime,1002,01/06/2014 17:48:32
soki,539,01/06/2014 17:48:52
gajimbi,319,01/06/2014 18:02:36
jifi,881,01/06/2014 18:02:55
jimbi,879,01/06/2014 18:03:12
jaru,863,01/06/2014 18:03:35
jang,846,01/06/2014 18:05:00
dosimbi,712,01/06/2014 18:07:58
wambi,1096,01/06/2014 18:08:56
wecembi,1109,01/06/2014 18:09:22
halaha,352,01/06/2014 19:27:09
sekiyen,509,01/06/2014 19:28:27
hanciki,344,01/06/2014 19:35:30
geren,970,01/06/2014 19:35:53
elbimbi,114,01/06/2014 19:36:23
dahabumbi,673,01/06/2014 19:37:04
hvrgi,390,01/06/2014 19:37:51
hada,351,01/06/2014 19:39:01
arambi,85,01/06/2014 19:39:56
cooha,835,01/06/2014 19:41:06
kambi,293,01/06/2014 19:41:59
afambi,92,01/06/2014 19:42:23
bahakv,412,01/06/2014 19:43:03
bederembi,424,01/06/2014 19:45:09
eshen,106,01/06/2014 19:46:26
emhun,126,01/06/2014 19:47:26
ejelembi,128,01/06/2014 19:47:46
oho,187,01/06/2014 19:48:11
ombi,192,01/06/2014 19:48:56
ede,109,01/06/2014 19:49:24
isinjimbi,156,01/06/2014 19:49:52
bubombi,462,01/06/2014 19:52:48
gukumbi,1001,01/06/2014 19:53:29
fulahvn,1080,01/06/2014 19:54:05
honin,367,01/06/2014 19:54:55
bolobi,454,01/06/2014 19:55:19
necihiyembi,259,01/06/2014 19:55:50
abkai,44,12/06/2014 21:54:32


dict.py
# -*- coding: utf-8 -*-
##A script for querying words in 《滿漢大詞典》. Page number will be returned, and the page in dictionary pdf file will be opend. Searching history is stored in history.txt.
##By Yize
##files used by this script: numlist.csv, settings.ini, history.txt.
##dictmanchu_1爲《滿漢大詞典》所用字母經處理後對應的詞序。

dictmanchu_1 = {"a":1, "e":2, "i":3, "o":4, "u":5, "v":6, "n":7, "k":8, "g":9, "h":10, "b":11, "p":12, "s":13, "x":14, "t":15, "d":16, "l":17, "m":18, "c":19, "j":20, "y":21, "K":22, "G":23, "H":24, "r":25, "f":26, "w":27, "T":28, "D":29, "Z":30, "S":31, "C":32, "J":33, "Q":34}
def preprocess_1(word):
    """Prepare a word for function converttonum."""
    
    ##replace consonants for Chinese loan words with single letter.
    word = word.replace("k'","K")
    word = word.replace("g'","G")
    word = word.replace("h'","H")
    word = word.replace("ts'","T")
    word = word.replace("dz","D")
    word = word.replace("z","Z")
    word = word.replace("sy","S")
    word = word.replace("c'y","C")
    word = word.replace("jy","J")
    word = word.replace("ts","Q")
    
    ##identify gender for letter "k", "g", and "h".
    feminine = ["e", "i", "u"]
    vowel = ["a", "e", "i", "o", "u", "v"]
    i = 0
    while i < len(word):
        if word[i] == "k": ##letter "k"
            if i < len(word) - 1: ##not final k
                if word[i+1] in feminine:
                    word = word[:i] + "K" + word[i+1:]
                kbcspecial = ["Ku", "Gu", "Hu", "hv"] ##special cases precede "k" when the "k" is followed by a consonant.
                if (word[i+1] not in vowel) and (word[i-1] == "e" or (i >= 2 and word[i-2:i] in kbcspecial)): ##"k" after "ku", "gu", "hu", "hv" ,or "e" when followed by a consonant. 
                    word = word[:i] + "K" + word[i+1:]
            elif i == len(word) - 1: ##final "k"
                if word[i-1] == "i" or (word[i-1] == "e" and word[i-2] != "t"): ##final "k" preceded by "i" or "e", but not "te"
                    word = word[:i] + "K"
        if word[i] == "g": ##letter "g"
            if i < len(word) - 1: ##not final "g"
                if word[i+1] in feminine: 
                    word = word[:i] + "G" + word[i+1:]
                elif word[i+1] not in vowel and word[i-1] == "n": #letter "g" in "ng+consonant" case
                    word = word[:i] + "G" + word[i+1:]
            elif i == len(word) - 1 and word[i-1] == "n": ##final "g", i.e. "ng" at the final part of a word.
                word = word[:i] + "G"
        if word[i] == "h" and word[i+1] in feminine: ##letter "h"
            word = word[:i] + "H" + word[i+1:]
        i += 1
    
    return word
    
def converttonum(word,index=6,**dict):
    """Convert a word to number."""
    m_notation = len(dict)
    wordnum = 0
    ##convert
    for letter in word:
        ##ilegal character is treated as 0.
        if letter in dict:
            wordnum += dict[letter] * pow(m_notation,index)
        index -= 1
    return wordnum

import csv
import time
import subprocess
import ConfigParser

##read configures from settings.ini
config = ConfigParser.SafeConfigParser({'openpdf': 0, 'savehistory': 1, 'command': '"C:\Program Files (x86)\Foxit Software\Foxit Reader\Foxit Reader.exe" C:\Users\adminadmin\Documents\課程\满语初程\電子詞典製作\dict_python\manchudict.pdf /A page='})
config.read('settings.ini')
openpdf = config.get('Section1','openpdf')
cmd = config.get('Section1','command')
savehistory = config.get('Section2','savehistory')

##read numlist.csv to start searching.
print 'Search Manhan Dacidian:'
with open('numlist.csv', 'rb') as f:
    reader = csv.reader(f)
    ##while loop for repeated searching.
    while True:
        print 'Please input a word (press Enter to exit):'
        searchword = raw_input()
        searchwordnum = converttonum(preprocess_1(searchword),**dictmanchu_1)
        if searchwordnum == 0:
            exit()
        print 'Word represented by number:', searchwordnum
        ##search
        f.seek(0)
        pagenum = 0
        for row in reader: 
            if float(row[0]) > searchwordnum:
                pagenum = 16 + int(row[1]) - 1
                break
        print 'Word', '"', searchword, '"', 'is most probably on page', pagenum, '(page', pagenum - 16, 'by dictionary page).'
        ##show searching result.
        if pagenum != 0: ##word is found.
            ##write record (word, page number, and system time) to history.txt.
            if savehistory == '1':
                with open('history.txt', 'a') as fhistory: 
                    historyrecord = searchword + ',' + str(pagenum) + ',' + time.strftime("%d/%m/%Y")+ ' ' + time.strftime("%H:%M:%S") + '\n'
                    fhistory.write(historyrecord)
            ##open page in dictionary pdf file.
            if openpdf == '1':
                arg = cmd + str(pagenum)
                subprocess.call(arg)
        else: ##word is not found.
            print 'Word', searchword, 'is not found.'
